File: Core/src/main/java/com/inverce/mod/core/collections/WeakArrayList.java
package com.inverce.mod.core.collections;

import android.support.annotation.NonNull;
import android.support.annotation.Nullable;

import java.lang.ref.WeakReference;
import java.util.AbstractList;
import java.util.ArrayList;
import java.util.Iterator;

/**
 * Implementation of arrayList with values stored in weak references
 * @param <T>
 */
public class WeakArrayList<T> extends AbstractList<T> {
    @NonNull
    protected ArrayList<WeakReference<T>> impl = new ArrayList<>(1);

    @Override @Nullable
    public T get(int index) {
        return impl.get(index).get();
    }

    @Override
    public int size() {
        return impl.size();
    }

    @Nullable
    @Override
    public T set(int index, T element) {
        WeakReference<T> old = impl.set(index, new WeakReference<>(element));
        return old != null ? old.get() : null;
    }

    @Override
    public void add(int index, T element) {
        impl.add(index, new WeakReference<>(element));
    }

    @Nullable
    @Override
    public T remove(int index) {
        WeakReference<T> old = impl.remove(index);
        return old != null ? old.get() : null;
    }

    public int clearEmptyReferences() {
        int cleared = 0;
        Iterator<T> it = iterator();
        while (it.hasNext()) {
            if (it.next() == null) {
                it.remove();
                cleared++;
            }
        }
        return cleared;
    }
}

File: Core/src/main/java/com/inverce/mod/v2/core/collection/WeakArrayList.kt
package com.inverce.mod.v2.core.collection

import java.lang.ref.WeakReference
import java.util.*

/**
 * Implementation of arrayList with values stored in weak references
 * @param <T>
</T> */
class WeakArrayList<T> : kotlin.collections.AbstractMutableList<T?>() {

    override val size: Int
        get() = impl.size

    override fun add(index: Int, element: T?) {
        impl.add(index, WeakReference(element))
    }

    override fun removeAt(index: Int): T? {
        val old = impl.removeAt(index)
        return old?.get()
    }

    override fun set(index: Int, element: T?): T? {
        val old = impl.set(index, WeakReference(element))
        return old?.get()
    }

    protected val impl = ArrayList<WeakReference<T?>>(1)

    override fun get(index: Int): T? {
        return impl[index]?.get()
    }

    fun clearEmptyReferences(): Int {
        var cleared = 0
        val it = iterator()
        while (it.hasNext()) {
            if (it.next() == null) {
                it.remove()
                cleared++
            }
        }
        return cleared
    }
}